### PR TITLE
fix(image): #MZI-212 fix external image in mail body

### PR DESCRIPTION
**Pull request by Alex-Delaveau**
## Describe your changes
Fix issue where images in mail body from external senders where not displayed in zimbra but were displayed in expert mode

Now when the user try to read the mail we retrieve images from zimbra rest api, encode them in base64 and replace them in the mail body.

There is 2 new conf needed : 
- **$zimbraCidPattern** : A regex to match a content-id, cid are footprint of image from external mail client. they are used as the source of the image but are not recognize by our implementation of zimbra. That's why we replace them by base64 images retrieved with zimbra rest api.
- **$zimbraDocumentEndpoint** : the endpoint to retrieve image from zimbra server
The value of these 2 new confs are :

zimbraCidPattern='cid:(.*?)\"'
zimbraDocumentEndpoint='/service/home/~/?auth=co&loc=fr_FR&id=' 
## Checklist tests
Send a mail from another client (Outlook, Thunderbird...) with a image in body (Copy/Paste it into the mail not attachement). Then in Zimbra check if the image is displayed correctly.  
## Issue ticket number and link
[MZI-212](https://jira.support-ent.fr/browse/MZI-212)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 1]
```diff
@@ -336,10 +356,97 @@ private void transformMessageZimbraToFront(JsonObject msgZimbra, JsonObject msgF
...
      * @param multiparts Array of multipart structure
      */
-    private void processMessageMultipart(JsonObject msgFront, JsonArray multiparts) {
+    private Future<Void> processMessageMultipart(JsonObject msgFront, JsonArray multiparts, UserInfos user) {
+        Promise<Void> promise = Promise.promise();
         Multipart mparts = new Multipart(msgFront.getString(FrontConstants.MESSAGE_ID), multiparts);
+
         msgFront.put(MESSAGE_BODY, mparts.getBody());
-        msgFront.put(MESSAGE_ATTACHMENTS, mparts.getAttachmentsJson());
+        replaceCidToUrl(mparts, msgFront, user).onComplete(ar -> {
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: .onSuccess(...)
.onFailure(...)

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 2]
```diff
@@ -336,10 +356,97 @@ private void transformMessageZimbraToFront(JsonObject msgZimbra, JsonObject msgF
...
+            if (ar.succeeded()) {
+                JsonObject updatedMsgFront = ar.result();
+                updatedMsgFront.put(MESSAGE_ATTACHMENTS, mparts.getAttachmentsJson());
+                msgFront.mergeIn(updatedMsgFront);
+                promise.complete();
+            } else {
+                // handle error
+                Throwable error = ar.cause();
+                log.error("Zimbra@processMessageMultipart : error processing multiparts: " + error.getMessage());
+                promise.fail(error);
```
Comment (qpic-cgi) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: ne remonte pas la stacktrace, remonte ton message plutôt:
```suggestion
                String message = String.format("[Zimbra@%s::processMessageMultipart ] Fail to store " +
                                            "permissions preferences to session. %s", this.getClass().getSimpleName()));
                log.error(String.format("%s %s", message, error.getMessage());
                promise.fail(message);
```

File: zimbra/src/main/java/fr/openent/zimbra/service/data/HttpService.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    Vertx vertx;
+    private static final Logger log = LoggerFactory.getLogger(MessageService.class);
+
+    public HttpService(Vertx vertx) {
+        this.vertx = vertx;
+    }
+
+    public Future<Buffer> getRequest(String url, HashMap<String, String> header) {
+        Promise<Buffer> promise = Promise.promise();
+        HttpClient httpClient = HttpClientHelper.createHttpClient(vertx);
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/service/data/HttpService.java]: Pourquoi ne pas utiliser WebClient ?

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 4]
```diff
@@ -60,17 +62,27 @@ public class MessageService {
...
-                          DbMailService dbMailService, UserService userService, SynchroUserService synchroUserService) {
+                          DbMailService dbMailService, UserService userService, SynchroUserService synchroUserService, HttpService httpService) {
         this.soapService = soapService;
         this.folderService = folderService;
         this.dbMailService = dbMailService;
         this.userService = userService;
         this.groupService = new GroupService(soapService, dbMailService, synchroUserService);
+        this.httpService = httpService;
+        if(Zimbra.appConfig != null) {
+            this.zimbraUrlGetImage = Zimbra.appConfig.getZimbraUri() + "/service/home/~/?auth=co&loc=fr_FR&id=";
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: D'où vient cet endpoint?

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 5]
```diff
@@ -317,16 +329,24 @@ private void transformMessageZimbraToFront(JsonObject msgZimbra, JsonObject msgF
...
-            processMessageMultipart(msgFront, multiparts);
-        }
-
 
-        translateMaillistToUidlist(msgFront, zimbraMails, addressMap,
-                msgZimbra.getString(MSG_FLAGS, "").contains(MSG_FLAG_NOTIFICATIONSENT), result);
+            processMessageMultipart(msgFront, multiparts, user).onComplete(ar -> {
+                if (ar.failed()) {
+                    Throwable error = ar.cause();
+                    log.error("Zimbra@transformMessageZimbraToFront : error transforming messages: " + error.getMessage());
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: Nécessaire de faire en 2 lignes ?
Tu peux ne pas utiliser ar.cause().getMessage() directement ?

Et malgré la "gestion" d'erreur, tu continues l'instruction

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 6]
```diff
@@ -336,10 +356,97 @@ private void transformMessageZimbraToFront(JsonObject msgZimbra, JsonObject msgF
...
+        replaceCidToUrl(mparts, msgFront, user).onComplete(ar -> {
+            if (ar.succeeded()) {
+                JsonObject updatedMsgFront = ar.result();
+                updatedMsgFront.put(MESSAGE_ATTACHMENTS, mparts.getAttachmentsJson());
+                msgFront.mergeIn(updatedMsgFront);
+                promise.complete();
+            } else {
+                // handle error
+                Throwable error = ar.cause();
+                log.error("Zimbra@processMessageMultipart : error processing multiparts: " + error.getMessage());
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: Tu peux faire directement en ar.cause().getMessage()

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 7]
```diff
@@ -336,10 +356,97 @@ private void transformMessageZimbraToFront(JsonObject msgZimbra, JsonObject msgF
...
+                promise.fail(error);
+            }
+        });
+        return promise.future();
+    }
+    private Future<JsonObject> replaceCidToUrl(Multipart  mparts, JsonObject msgFront, UserInfos user) {
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: Je veux bien que tu détailles ton opération et le cid qu'est-ce que tu fais exactement ?

**Review by JinDouang — CHANGES_REQUESTED** [review]

**Review by qpic-cgi — CHANGES_REQUESTED** [review]

File: zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        try {
+            Pattern pattern = Pattern.compile(Zimbra.appConfig.getZimbraDocumentConfig().getString("zimbraCidPattern"));
+            Matcher matcher = pattern.matcher(message.getString(MESSAGE_BODY));
+            List<String> cids = new ArrayList<>();
+            while (matcher.find()) {
+                cids.add(matcher.group(1));
+            }
+            return cids;
+        } catch (NullPointerException npe){
+            return new ArrayList<>();
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java:35]: tu devrais rajouter log erreur

File: zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+public class CidHelper {
+
+    private CidHelper() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    public static List<String> getMessageCids(JsonObject message){
+        try {
+            Pattern pattern = Pattern.compile(Zimbra.appConfig.getZimbraDocumentConfig().getString("zimbraCidPattern"));
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java]: const ?

File: zimbra/src/main/java/fr/openent/zimbra/helper/ConfigManager.java [hunk 10]
```diff
@@ -98,6 +98,8 @@ public class ConfigManager {
 
     private SlackConfiguration slackConfiguration;
 
+    private final JsonObject zimbraDocumentConfig;
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/helper/ConfigManager.java]: tu pouvais pas le modéliser ?

File: zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java (hunk 9 above)
Comment (JinDouang) [reply zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java]: tu pouvais pas le modéliser ?

File: zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                cids.add(matcher.group(1));
+            }
+            return cids;
+        } catch (NullPointerException npe){
+            return new ArrayList<>();
+        }
+    }
+
+    public static Map<String, String> mapCidToImageUrl(String messageId, List<String> cids){
+        String zimbraUrlGetImage = Zimbra.appConfig.getZimbraUri() + Zimbra.appConfig.getZimbraDocumentConfig().getString("zimbraDocumentEndpoint");;
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java]: tu pouvais pas le modéliser ?

File: zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+        HashMap<String, String> mapCidToImageUrl = new HashMap<>();
+        int partCount = 2;
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/helper/CidHelper.java:43]: pourquoi 2 ?

peux-tu éclaircir ton raisonnement dans ta boucle ?

File: zimbra/src/main/java/fr/openent/zimbra/helper/ConfigManager.java [hunk 13]
```diff
@@ -175,6 +177,9 @@ public ConfigManager(JsonObject config) {
         this.circuitBreakerOptions = new CircuitBreakerOptions(config.getJsonObject("circuit-breaker", new JsonObject()));
         JsonObject slackConfig = config.getJsonObject("slack", new JsonObject());
         this.slackConfiguration = new SlackConfiguration(slackConfig.getString("api-uri", ""), slackConfig.getString("api-token", ""), slackConfig.getString("channel", ""), slackConfig.getString("bot-username", ""), config.getString("host", ""));
+        this.zimbraDocumentConfig = config.getJsonObject("zimbra-document-config", new JsonObject()
```
Comment (JinDouang) [inline zimbra/src/main/java/fr/openent/zimbra/helper/ConfigManager.java]: tu pouvais pas le modéliser ?

File: zimbra/src/test/java/fr/openent/zimbra/helper/CidHelperTest.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+@RunWith(VertxUnitRunner.class)
+public class CidHelperTest {
+
+    @Before
+    public void before(){
+        ConfigManager cm = new ConfigManager(new JsonObject());
+        Zimbra.appConfig = cm;
+    }
+    @Test
+    public void testGetCids (TestContext ctx) {
```
Comment (JinDouang) [inline zimbra/src/test/java/fr/openent/zimbra/helper/CidHelperTest.java]: précise ce que ton TU est censé faire en terme de nommage

**Review by JinDouang — CHANGES_REQUESTED** [review]

**Review by JinDouang — APPROVED** [review]

**Pull request merged**